pkg/resources/kafka/configmap.go

```diff
@@ -37,17 +41,38 @@ func (r *Reconciler) configMapPod(broker banzaicloudv1alpha1.BrokerConfig, loadB
 			fmt.Sprintf("cruise.control.metrics.reporter.bootstrap.servers=%s\n", strings.Join(getInternalListeners(r.KafkaCluster.Spec.ListenersConfig.InternalListeners, broker, r.KafkaCluster.Namespace, r.KafkaCluster.Name, r.KafkaCluster.Spec.HeadlessServiceEnabled), ",")) +
 			fmt.Sprintf("broker.id=%d\n", broker.Id) +
 			generateStorageConfig(broker.StorageConfigs) +
-			generateAdvertisedListenerConfig(broker, r.KafkaCluster.Spec.ListenersConfig, loadBalancerIP, r.KafkaCluster.Namespace, r.KafkaCluster.Name, r.KafkaCluster.Spec.HeadlessServiceEnabled) +
+			generateAdvertisedListenerConfig(r.Client, broker, r.KafkaCluster.Spec.ListenersConfig, loadBalancerIP, r.KafkaCluster.Namespace, r.KafkaCluster.Name, r.KafkaCluster.Spec.HeadlessServiceEnabled) +
 			broker.Config},
 	}
 }
 
-func generateAdvertisedListenerConfig(broker banzaicloudv1alpha1.BrokerConfig, l banzaicloudv1alpha1.ListenersConfig, loadBalancerIP, namespace, crName string, headlessServiceEnabled bool) string {
+func generateAdvertisedListenerConfig(cli client.Client, broker banzaicloudv1alpha1.BrokerConfig, l banzaicloudv1alpha1.ListenersConfig, loadBalancerIP, namespace, crName string, headlessServiceEnabled bool) string {
 	advertisedListenerConfig := []string{}
-	for _, eListener := range l.ExternalListeners {
-		advertisedListenerConfig = append(advertisedListenerConfig,
-			fmt.Sprintf("%s://%s:%d", strings.ToUpper(eListener.Name), loadBalancerIP, eListener.ExternalStartingPort+broker.Id))
+
+	if loadBalancerIP != "" {
+		for _, eListener := range l.ExternalListeners {
+			advertisedListenerConfig = append(advertisedListenerConfig,
+				fmt.Sprintf("%s://%s:%d", strings.ToUpper(eListener.Name), loadBalancerIP, eListener.ExternalStartingPort+broker.Id))
+		}
+	} else {
+		var svc corev1.Service
+		var pods corev1.PodList
+		if err := cli.List(context.TODO(), &pods,
+			client.MatchingLabels{"kafka_cr": crName, "brokerId": string(broker.Id)}); err != nil {
+			log.Error(err)
+		}
+		pod := pods.Items[0]
+		if err := cli.Get(context.TODO(), types.NamespacedName{Namespace: namespace,
+			Name: fmt.Sprintf("%s-%d-svc", crName, broker.Id)}, &svc); err != nil {
+			log.Error(err)
+		}
+
+		for _, eListener := range l.ExternalListeners {
+			advertisedListenerConfig = append(advertisedListenerConfig,
+				fmt.Sprintf("%s://%s:%d", strings.ToUpper(eListener.Name), pod.Status.HostIP, svc.Spec.Ports[0].NodePort))
```

Comment:
Using `pod.Status.HostIP` as an external access address will not work in every case.
For example.: on an Amazon cluster the `pod.Status.HostIP` will contain an address which is not a public IP.
Unfortunately the public IP of the Node will not be available in the Pod. 
What you need to do is instead wait until the pod is scheduled to a node, get the `nodeName` field from the Pod's spec, get the given node from Kubernetes. 
The node has something called `addresses` in it's status which contains the required information. https://kubernetes.io/docs/concepts/architecture/nodes/#addresses.

I recommend to use a logic like we did in case of the rack awareness feature.
Basically we delayed the ConfigMap creation on the broker until we have all required information. Since Kubernetes schedules the Pod on a suitable Node before attaching the ConfigMap, which means you have all required information to generate the config.

Reply:
Thanks for your advise. I will re-check it.

Reply:
Sorry for the delayed reply. If that is your case I think we should not hard code anything. I recommend to let the user to make the decision. There are multiple `NodeAddressType` available in the above mentioned status field, I think we should redesign the CRD to allow to set the address type.